### ISSUE TITLE
Add Magnitude>>between:and:

### ISSUE THREAD

**Issue by hhzl** [amber-smalltalk/amber#1181]
A method which Pharo has

    between: min and: max 
    	"Answer whether the receiver is less than or equal to the argument, max, 
    	and greater than or equal to the argument, min."

    	^self >= min and: [self <= max]

**Comment by hhzl** [comment]
There is no class ``Magnitude`` in Amber but a class ``Number`` which is a  subclass of ``Magnitude`` in ``Pharo``. So let's put it there into category ``testing``.

**Comment by herby** [comment]
You can add `Magnitude`. Not directly, though, as `Number` as well as `Collection` are critical classes and thus Amber probably fails while trying to reparent them (#382), but you can add Magnitude class in boot.js, in that same place where `Number` and `Array` are wrapped and where basic `Collection` hierarchy is set up. Just insert one more call and fix the parameters in a bunch of others. Then reload, and save the package where the class definition was added; it is saved and from that point you can continue inside IDE.

**Comment by hhzl** [comment]
You mean this function

````
function SmalltalkInitBrik(brikz, st) {
````
in ``boot.js`` in the ``support`` folder?

````
        this.__init__ = function () {
            st.addPackage("Kernel-Methods");
            st.wrapClassName("Number", "Kernel-Objects", Number, globals.Object);
            st.wrapClassName("BlockClosure", "Kernel-Methods", Function, globals.Object);
            st.wrapClassName("Boolean", "Kernel-Objects", Boolean, globals.Object);
            st.wrapClassName("Date", "Kernel-Objects", Date, globals.Object);

            st.addPackage("Kernel-Collections");
            st.addClass("Collection", globals.Object, null, "Kernel-Collections");
            st.addClass("IndexableCollection", globals.Collection, null, "Kernel-Collections");
            st.addClass("SequenceableCollection", globals.IndexableCollection, null, "Kernel-Collections");
            st.addClass("CharacterArray", globals.SequenceableCollection, null, "Kernel-Collections");
            st.wrapClassName("String", "Kernel-Collections", String, globals.CharacterArray);
            st.wrapClassName("Array", "Kernel-Collections", Array, globals.SequenceableCollection);
            st.wrapClassName("RegularExpression", "Kernel-Collections", RegExp, globals.Object);

            st.addPackage("Kernel-Exceptions");
            st.wrapClassName("Error", "Kernel-Exceptions", Error, globals.Object);

            /* Alias definitions */

            st.alias(globals.Array, "OrderedCollection");
            st.alias(globals.Date, "Time");

        };
````

And I suggest to remove the ``entry level`` label

**Comment by herby** [comment]
Yeah, this one. Actually, adding between:and: is entry level, adding the Magnitude class isn't. May be adding Magnitude class itself can be another issue (or it just can be done separately, and once it's done, this actually remains entry level) and pulling up common API to Magnitude class can be yet another issue.

Adding the class just matter adding one `addClass` call and fixing a few others so Magnitude is used as the superclass of `Number` and `Collection`.

**Comment by herby** [comment]
But yeah, I'll remove entry level until Magnitude is added.

**Comment by hhzl** [comment]
To add class ``Magnitude`` I need a new line in function ``SmalltalkInitBrik(brikz, st)``
    
    st.addClass("Magnitude", globals.Object, null, "Kernel-Objects");  // new

The parameters of ``addClass`` are ``(className, superclass, iVarNames, pkgName)``.

Then the classes ``Number`` and ``Date`` have to be made subclasses of ``Magnitude``.
This results in a change from the previous superclass ``Object`` to ``Magnitude`` in the following to existing lines

````
st.wrapClassName("Number", "Kernel-Objects", Number, globals.Magnitude); 
st.wrapClassName("Date", "Kernel-Objects", Date, globals.Magnitude);  
````


More context of the code in function ``function SmalltalkInitBrik(brikz, st)`` follows:

````
this.__init__ = function () {

            st.addClass("Magnitude", globals.Object, null, "Kernel-Objects"); 

            st.wrapClassName("Number", "Kernel-Objects", Number, globals.Magnitude);  
            st.wrapClassName("Date", "Kernel-Objects", Date, globals.Magnitude);   

            st.addPackage("Kernel-Methods");
            st.wrapClassName("BlockClosure", "Kernel-Methods", Function, globals.Object);
            st.wrapClassName("Boolean", "Kernel-Objects", Boolean, globals.Object);

            st.addPackage("Kernel-Collections");
            st.addClass("Collection", globals.Object, null, "Kernel-Collections");
            st.addClass("IndexableCollection", globals.Collection, null, "Kernel-Collections");
            st.addClass("SequenceableCollection", globals.IndexableCollection, null, "Kernel-Collections");
            st.addClass("CharacterArray", globals.SequenceableCollection, null, "Kernel-Collections");
            st.wrapClassName("String", "Kernel-Collections", String, globals.CharacterArray);
            st.wrapClassName("Array", "Kernel-Collections", Array, globals.SequenceableCollection);
            st.wrapClassName("RegularExpression", "Kernel-Collections", RegExp, globals.Object);

            st.addPackage("Kernel-Exceptions");
            st.wrapClassName("Error", "Kernel-Exceptions", Error, globals.Object);

            /* Alias definitions */

            st.alias(globals.Array, "OrderedCollection");
            st.alias(globals.Date, "Time");

        };
````

Is this all to insert the superclass ``Magnitude`` on the js level? There is no class ``Character``. 

Methods are added later in the IDE. 

Note: Pharo 4.0 has
````
Magnitude subclasses  "{Number. Character. LookupKey. DateAndTime. Duration. Timespan. Time. MessageTally. WeakKeyAssociation. MetacelloVersion. MetacelloVersionNumber. MetacelloSemanticVersionNumber. TxBasicTextPosition}"
````

**Comment by hhzl** [comment]
If I do the above changed to ``boot.js`` Amber does not come up with the IDE selection dialog and I get the following error message in the browser console

````
mutating the [[Prototype]] of an object will cause your code to run very slowly; instead create the object with the correct initial [[Prototype]] value using Object.create frameworks-3241a40a58a82e21daef3dd3cdca01bde189158793c1b6f9193fff2b5293cd1d.js:34
TypeError: $recv(...).__eq is not a function Kernel-Objects.js:457
````

**Comment by herby** [comment]
Ah, yes, in Kernel-Objects.js there is another call to addClass with different superclass so it tries to do some more work than needed. You must edit the file as well to have Magnitude there so it sees no change in superclass.

**Comment by hhzl** [comment]
I did the following changes to ``Kernel-Objects.js``

Line 2218

changed:
``$core.addClass('Date', $globals.Object, [], 'Kernel-Objects');``
to:
``$core.addClass('Date', $globals.Magnitude, [], 'Kernel-Objects');``


Line 3263
changed
``$core.addClass('Number', $globals.Object, [], 'Kernel-Objects');``
to:
``$core.addClass('Number', $globals.Magnitude, [], 'Kernel-Objects');``

this time the browser console remained empty which makes it hard to find out what is going on.

Adding 

``$core.addClass('Magnitude', $globals.Object, [], 'Kernel-Objects');``
in front of line 2218 did not help.

**Comment by hhzl** [comment]
Correction: there was no error message in the browser console because it seemed to have worked.

**Comment by herby** [comment]
It seems it worked means "IDE started, I could commit the Kernel-Objects package which now contains evertyhing in order"?

**Comment by hhzl** [comment]
I could commit the package ``Kernel-Objects`` after having added the method ``#between:and`` to class ``Magnitude``. Magnitude has subclasses ``Number`` and ``Date``. This was done in Helios.

New commit: https://github.com/hhzl/amber/tree/issue-1181

**Comment by hhzl** [comment]
Kernel and Compiler tests in SUnit (Helios) 523 passes out of 523

![testing_protocol_of_magnitude_in_pharo_4 0_2015-07-02](https://cloud.githubusercontent.com/assets/1112023/8489928/bfa55ece-2110-11e5-9927-ccefc2d722c9.png)

What should be done with the rest of the protocol of Magnitude?

Pharo 4.0

    Magnitude selectors 
    "#(#hash #> #max:in: #'>=' #min: #< #max: #min:max:in: 
     #min:max: #= #between:and: 
    #compareWith:ifLesser:ifEqual:ifGreater: #'<=')"

**Comment by herby** [comment]
If Tests pass, issue a PR. If not yet present, rebase with reword or commit --amend so that the commit actually adding #between:and: includes "Fix #1181" or "Fixes #1181".

**Comment by herby** [comment]
As for the rest of the protocol, you can get it in (except hash, of course).
